chain/bitcoind.go

```diff
@@ -0,0 +1,1122 @@
+}
+
+// SetStartTime is a non-interface method to set the birthday of the wallet
+// using this object. Since only a single rescan at a time is currently
+// supported, only one birthday needs to be set. This does not fully restart a
+// running rescan, so should not be used to update a rescan while it is running.
+// TODO: When factoring out to multiple rescans per bitcoind client, add a
+// birthday per client.
+func (c *BitcoindClient) SetStartTime(startTime time.Time) {
+	c.clientMtx.Lock()
+	defer c.clientMtx.Unlock()
+
+	c.startTime = startTime
+}
+
+// BlockStamp returns the latest block notified by the client, or an error
+// if the client has been shut down.
+func (c *BitcoindClient) BlockStamp() (*waddrmgr.BlockStamp, error) {
+	select {
+	case bs := <-c.currentBlock:
+		return bs, nil
+	case <-c.quit:
+		return nil, errors.New("disconnected")
+	}
+}
+
+func (c *BitcoindClient) onClientConnect() {
+	select {
+	case c.enqueueNotification <- ClientConnected{}:
+	case <-c.quit:
+	}
+}
+
+func (c *BitcoindClient) onBlockConnected(hash *chainhash.Hash, height int32, time time.Time) {
+	select {
+	case c.enqueueNotification <- BlockConnected{
+		Block: wtxmgr.Block{
+			Hash:   *hash,
+			Height: height,
+		},
+		Time: time,
+	}:
+	case <-c.quit:
+	}
+}
+
+func (c *BitcoindClient) onFilteredBlockConnected(height int32,
+	header *wire.BlockHeader, relevantTxs []*wtxmgr.TxRecord) {
+	select {
+	case c.enqueueNotification <- FilteredBlockConnected{
+		Block: &wtxmgr.BlockMeta{
+			Block: wtxmgr.Block{
+				Hash:   header.BlockHash(),
+				Height: height,
+			},
+			Time: header.Timestamp,
+		},
+		RelevantTxs: relevantTxs,
+	}:
+	case <-c.quit:
+	}
+}
+
+func (c *BitcoindClient) onBlockDisconnected(hash *chainhash.Hash, height int32, time time.Time) {
+	select {
+	case c.enqueueNotification <- BlockDisconnected{
+		Block: wtxmgr.Block{
+			Hash:   *hash,
+			Height: height,
+		},
+		Time: time,
+	}:
+	case <-c.quit:
+	}
+}
+
+func (c *BitcoindClient) onRelevantTx(rec *wtxmgr.TxRecord,
+	block *btcjson.BlockDetails) {
+	blk, err := parseBlock(block)
+	if err != nil {
+		// Log and drop improper notification.
+		log.Errorf("recvtx notification bad block: %v", err)
+		return
+	}
+
+	select {
+	case c.enqueueNotification <- RelevantTx{rec, blk}:
+	case <-c.quit:
+	}
+}
+
+func (c *BitcoindClient) onRescanProgress(hash *chainhash.Hash, height int32, blkTime time.Time) {
+	select {
+	case c.enqueueNotification <- &RescanProgress{hash, height, blkTime}:
+	case <-c.quit:
+	}
+}
+
+func (c *BitcoindClient) onRescanFinished(hash *chainhash.Hash, height int32, blkTime time.Time) {
+	log.Infof("Rescan finished at %d (%s)", height, hash)
+	select {
+	case c.enqueueNotification <- &RescanFinished{hash, height, blkTime}:
+	case <-c.quit:
+	}
+
+}
+
+// socketHandler reads events from the ZMQ socket, processes them as
+// appropriate, and queues them as btcd or neutrino would.
+func (c *BitcoindClient) socketHandler() {
+	defer c.wg.Done()
```

Comment:
This function as is, is a bit large. At the very least, the re-org logic can be extracted into a helper method. As bonus, this makes re-orgs easier to test in the future.